File: CICADA_FIGURES/plot_survival_curves_only.m
%% Plot Survival Curves with Confidence Intervals (No Swimmer Plots)
% This script loads precomputed bootstrap confidence bands and plots
% survival curves with 95% confidence intervals and RCT reference curves

clear all; close all; clc;

%% Load precomputed bootstrap confidence bands
load('bootstrap_confidence_bands.mat');
% This loads: S0h, S1h, s0_lower, s0_upper, s1_lower, s1_upper, 
%             s0_median, s1_median, t_grid, Nboot

%% Get reference curves from RCT data
T1 = readtable('trialData1.csv');
[s0_true, s1_true, t0_true, t1_true] = fcnPlotKM(T1);

%% Create figure
figure('Position', [100, 100, 800, 600]);
hold on;

% Time vector
t_grid = 0:2:168;  % 85 points

% Convert to column vectors for consistency
t_grid = t_grid(:);
s0_lower = s0_lower(:);
s0_upper = s0_upper(:);
s1_lower = s1_lower(:);
s1_upper = s1_upper(:);
s0_median = s0_median(:);
s1_median = s1_median(:);

%% Plot confidence bands (behind the lines)
% Untreated confidence band (blue)
fill([t_grid; flipud(t_grid)], ...
     [s0_lower; flipud(s0_upper)], ...
     [0.2, 0.4, 0.8], 'FaceAlpha', 0.3, 'EdgeColor', 'none', ...
     'DisplayName', 'Untreated 95% CI');

% Treated confidence band (red)
fill([t_grid; flipud(t_grid)], ...
     [s1_lower; flipud(s1_upper)], ...
     [0.8, 0.2, 0.2], 'FaceAlpha', 0.3, 'EdgeColor', 'none', ...
     'DisplayName', 'Treated 95% CI');

%% Plot main curves
% True/RCT reference curves (dashed)
plot(t0_true, s0_true, 'b--', 'LineWidth', 2.5, 'DisplayName', 'Untreated (RCT)');
plot(t1_true, s1_true, 'r--', 'LineWidth', 2.5, 'DisplayName', 'Treated (RCT)');

% Estimated curves from bootstrap median (solid)
plot(t_grid, s0_median, 'b-', 'LineWidth', 2.5, 'DisplayName', 'Untreated (Estimated)');
plot(t_grid, s1_median, 'r-', 'LineWidth', 2.5, 'DisplayName', 'Treated (Estimated)');

%% Formatting
xlabel('Time (hours)', 'FontSize', 14);
ylabel('Survival Probability', 'FontSize', 14);
title('Survival Curves with 95% Bootstrap Confidence Intervals', 'FontSize', 16);
legend('Location', 'southwest', 'FontSize', 11);
grid on;
set(gca, 'FontSize', 12);
xlim([0, 168]);
ylim([0, 1]);
set(gcf, 'Color', 'w');

% Set x-axis ticks every 24 hours
xticks(0:24:168);
xticklabels(0:24:168);

%% Display summary statistics
fprintf('\n=== Survival Analysis Summary ===\n');
fprintf('Bootstrap samples: %d\n\n', Nboot);

% Convert back to row vectors for analysis
s0_lower = s0_lower';
s0_upper = s0_upper';
s1_lower = s1_lower';
s1_upper = s1_upper';
s0_median = s0_median';
s1_median = s1_median';

% Survival at key time points
fprintf('Survival Probabilities:\n');
fprintf('-------------------------------------------\n');
fprintf('Time (h) | Untreated         | Treated\n');
fprintf('         | Median [95%% CI]   | Median [95%% CI]\n');
fprintf('-------------------------------------------\n');

key_times = [24, 48, 72, 96, 120, 144, 168];
for kt = key_times
    [~, idx] = min(abs(t_grid' - kt));
    fprintf('%8d | %.1f%% [%.1f-%.1f] | %.1f%% [%.1f-%.1f]\n', ...
        kt, ...
        s0_median(idx)*100, s0_lower(idx)*100, s0_upper(idx)*100, ...
        s1_median(idx)*100, s1_lower(idx)*100, s1_upper(idx)*100);
end

% Treatment effect at end of study
fprintf('\n-------------------------------------------\n');
treatment_effect_median = s1_median(end) - s0_median(end);
fprintf('Treatment Effect at 168 hours:\n');
fprintf('  Absolute Risk Reduction: %.1f%%\n', treatment_effect_median*100);
fprintf('  Number Needed to Treat: %.1f\n', 1/treatment_effect_median);

%% Save figure
print('Fig_survival_curves_only.pdf', '-dpdf', '-r300');
fprintf('\nFigure saved as: Fig_survival_curves_only.pdf\n');

File: CICADA_FIGURES_PY/plot_survival_curves_only.py
"""
Python port of CICADA_FIGURES/plot_survival_curves_only.m

Diagnostic utility: plots survival curves with 95% bootstrap CIs, overlaid
with RCT reference KM curves. Output: Fig_survival_curves_only.pdf.
"""

import matplotlib

matplotlib.use("Agg")

import os
import sys

HERE = os.path.dirname(os.path.abspath(__file__))
REPO_ROOT = os.path.abspath(os.path.join(HERE, ".."))
sys.path.insert(0, os.path.join(REPO_ROOT, "python"))

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from scipy.io import loadmat

np.random.seed(0)

from fcnPlotKM import fcnPlotKM  # type: ignore


def _load_mat(path):
    try:
        return loadmat(path)
    except NotImplementedError:
        import h5py

        with h5py.File(path, "r") as h:
            return {k: np.array(h[k]) for k in h.keys()}


def main():
    boot_path = os.path.join(REPO_ROOT, "bootstrap_confidence_bands.mat")
    trial_path = os.path.join(REPO_ROOT, "trialData1.csv")

    if not os.path.exists(boot_path):
        print(f"[plot_survival_curves_only] Missing: {boot_path}. Exiting.")
        return 0
    if not os.path.exists(trial_path):
        print(f"[plot_survival_curves_only] Missing: {trial_path}. Exiting.")
        return 0

    d = _load_mat(boot_path)
    s0_lower = np.asarray(d["s0_lower"], dtype=float).ravel()
    s0_upper = np.asarray(d["s0_upper"], dtype=float).ravel()
    s1_lower = np.asarray(d["s1_lower"], dtype=float).ravel()
    s1_upper = np.asarray(d["s1_upper"], dtype=float).ravel()
    s0_median = np.asarray(d["s0_median"], dtype=float).ravel()
    s1_median = np.asarray(d["s1_median"], dtype=float).ravel()
    Nboot = int(np.asarray(d["Nboot"]).ravel()[0])

    T1 = pd.read_csv(trial_path)
    s0_true, s1_true, t0_true, t1_true = fcnPlotKM(T1)

    t_grid = np.arange(0, 168 + 2, 2)

    fig = plt.figure(figsize=(8, 6), facecolor="white")
    ax = fig.add_subplot(1, 1, 1)

    ax.fill_between(
        t_grid,
        s0_lower,
        s0_upper,
        facecolor=(0.2, 0.4, 0.8),
        alpha=0.3,
        edgecolor="none",
        label="Untreated 95% CI",
    )
    ax.fill_between(
        t_grid,
        s1_lower,
        s1_upper,
        facecolor=(0.8, 0.2, 0.2),
        alpha=0.3,
        edgecolor="none",
        label="Treated 95% CI",
    )
    ax.plot(t0_true, s0_true, "b--", linewidth=2.5, label="Untreated (RCT)")
    ax.plot(t1_true, s1_true, "r--", linewidth=2.5, label="Treated (RCT)")
    ax.plot(t_grid, s0_median, "b-", linewidth=2.5, label="Untreated (Estimated)")
    ax.plot(t_grid, s1_median, "r-", linewidth=2.5, label="Treated (Estimated)")

    ax.set_xlabel("Time (hours)", fontsize=14)
    ax.set_ylabel("Survival Probability", fontsize=14)
    ax.set_title(
        "Survival Curves with 95% Bootstrap Confidence Intervals", fontsize=16
    )
    ax.legend(loc="lower left", fontsize=11)
    ax.grid(True)
    ax.tick_params(labelsize=12)
    ax.set_xlim(0, 168)
    ax.set_ylim(0, 1)
    ax.set_xticks(np.arange(0, 169, 24))

    # Summary
    print("\n=== Survival Analysis Summary ===")
    print(f"Bootstrap samples: {Nboot}")
    print("\nSurvival Probabilities:")
    print("-------------------------------------------")
    print("Time (h) | Untreated         | Treated")
    print("         | Median [95% CI]   | Median [95% CI]")
    print("-------------------------------------------")
    for kt in [24, 48, 72, 96, 120, 144, 168]:
        idx = int(np.argmin(np.abs(t_grid - kt)))
        print(
            f"{kt:8d} | {s0_median[idx]*100:.1f}% [{s0_lower[idx]*100:.1f}-{s0_upper[idx]*100:.1f}] | "
            f"{s1_median[idx]*100:.1f}% [{s1_lower[idx]*100:.1f}-{s1_upper[idx]*100:.1f}]"
        )

    te = s1_median[-1] - s0_median[-1]
    print("\n-------------------------------------------")
    print("Treatment Effect at 168 hours:")
    print(f"  Absolute Risk Reduction: {te*100:.1f}%")
    if te != 0:
        print(f"  Number Needed to Treat: {1/te:.1f}")

    out_path = os.path.join(HERE, "Fig_survival_curves_only.pdf")
    fig.savefig(out_path, dpi=300)
    plt.close(fig)
    print(f"\nFigure saved as: {out_path}")
    return 0


if __name__ == "__main__":
    sys.exit(main())
